### DISCUSSION TITLE
Question: Why does this import work by default, but not with --import-mode=importlib

### DISCUSSION THREAD

**Discussion by LuminairPrime** [pytest-dev/pytest#9835] [Q&A]
I was testing --import-mode=importlib because I read that it will eventually be the default mode. 

I have test_x.py with:
```
import y
def test_x(): assert 1
```

And y.py with:
```
print("test?")
```
The output with a default pytest is a pass:

```
C:\pytest_experiment>pytest
========================================================== test session starts ==========================================================
platform win32 -- Python 3.10.1, pytest-7.1.1, pluggy-1.0.0
rootdir: C:\pytest_experiment
plugins: forked-1.4.0, monkeytype-1.1.0, xdist-2.5.0
collected 1 item

test_x.py .                                                                                                                        [100%]

=========================================================== 1 passed in 0.02s ===========================================================
```

And the output with a pytest --import-mode=importlib is a failed import:

```
C:\pytest_experiment>pytest --import-mode=importlib
========================================================== test session starts ==========================================================
platform win32 -- Python 3.10.1, pytest-7.1.1, pluggy-1.0.0
rootdir: C:\pytest_experiment
plugins: forked-1.4.0, monkeytype-1.1.0, xdist-2.5.0
collected 0 items / 1 error

================================================================ ERRORS =================================================================
______________________________________________________ ERROR collecting test_x.py _______________________________________________________
ImportError while importing test module 'C:\pytest_experiment\test_x.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
test_x.py:1: in <module>
    import y
E   ModuleNotFoundError: No module named 'y'
======================================================== short test summary info ========================================================
ERROR test_x.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=========================================================== 1 error in 0.11s ============================================================
```

What's up? Is that a bug, or do I need to do something else with --import-mode=importlib ?

**Comment by nicoddemus** [comment]
Hi @LuminairPrime, 

> I was testing --import-mode=importlib because I read that it will eventually be the default mode.

Just a note: I don't think that will be the case, because unfortunately it has its own issues as well.

> What's up? 

pytest needs to import `test_x.py` in order to collect and run the tests:

1. The default mode (`append`), will append the directory that contains `test_x.py` to `sys.path` and then call `__import__`.
2. The `importlib` mode uses a separate mechanism which doesn't require pytest to change `sys.path`.

The `append` mode works in that case because `sys.path` is updated to import `test_x.py`, which causes `y` to be importable.

**Comment by LuminairPrime** [comment]
Hmmm. In this case, the path that import-mode=importlib needs to find is equal to 1) the directory where pytest is run, and 2) the test file directory, and 3) the tested file directory. Shouldn't every import mode succeed at this?

Where do I need to place y.py for import-mode=importlib to find it?

**Comment by nicoddemus** [comment]
> In this case, the path that import-mode=importlib needs to find is equal to 1) the directory where pytest is run, and 2) the test file directory, and 3) the tested file directory. Shouldn't every import mode succeed at this?

1. Not really, just running `pytest` by itself doesn't change the `sys.path`. What changed the `sys.path` to the current directory was the fact that `test_x.py` was in the current directory, and `--import-mode=append` changes `sys.path` to import test files.
2. Correct.
3. Hmm this seems to be the same as 2, unless I'm misunderstanding.

Indeed import modes and test discovery is a somewhat complicated topic (because Python's import is complicated). We have a dedicated section in the docs dedicated to this topic: https://docs.pytest.org/en/7.1.x/explanation/pythonpath.html

> Where do I need to place y.py for import-mode=importlib to find it?

Usually the recommendation is using a virtual environment and installing your application there.

For quick one-off scripts or very simple projects, you can run `python -m pytest` instead of `pytest`, because the former adds the current directory to `sys.path` (that is normal Python behavior when executing `python -m`, not being specific to pytest).

**Comment by LuminairPrime** [comment]
For my personal use the default behavior of pytest works, and I can apply sys.path.append for extra distant directories. What I'm trying to get to the bottom of here is whether import-mode=importlib is broken. To ME it seems like it should work in this case.
